File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
 class Solution {
    public int lengthOfLongestSubstring(String s) {
        int ans = 0;
        HashMap<Character,Integer> map = new HashMap<>();
        
        int acq = 0;
        int rl = 0;
        
        for(;acq < s.length();acq++){
            char cur = s.charAt(acq);
            
            while(map.containsKey(cur)){
                map.remove(s.charAt(rl));
                rl++;
            }
            
            map.put(cur,1);
            
            ans = Math.max(ans,(acq - rl) + 1);
        }
        
        return ans;
    }
}

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        max_len = window_start = 0
        char_index_map = {}
        for window_end in range(len(s)):
            right_char = s[window_end]
            if right_char in char_index_map:
                window_start = max(window_start, char_index_map[right_char] + 1)
            char_index_map[right_char] = window_end
            max_len = max(max_len, window_end - window_start + 1)
        return max_len
            
                